Update costs

addysunn opened the pull request:
## Stuff not done:
- handle task deletion (the code is commented out cuz it needs to be udpate with our new ways of doing stuff)
- handle recipes for adding/removing input (this is on hold till I finish the below thing)

- One major thing I started Monday, but am still tinkering to get exactly right, is the fact that `update_parents_for_ingredient` has all this logic, but it all can be entirely be replaced by calling `update_children_after_batch_size_change_or_child_input_deleted` on each parent for the ingredient type. When this is finished, this fixes the bug we found on Friday's functionality review. With hindsight, I wish I could have realized this waaay long ago.

## Notable:
- All async functions are post_save, except deleting an input, which is pre_delete.

ScarpMetal committed:
Added comments to signals.py where Aditya should add his functions

addysunn committed:
propagate costs

- added remaining_worth column in task table
- one time command to propagate costs and remaining_worth
- function to propagate input changes in background using signals

addysunn committed:
cost changes - input delete

addysunn committed:
implement update cost

implement cost updates for batch_size change, input added/deleted

ishitap commented on ics/async_actions.py:
```diff
@@ -15,15 +19,197 @@ def update_task_descendents_flag_number(**kwargs):
+		if input_parent.remaining_worth is None:
+			parent_remaining_worth = 0
+		else:
+			parent_remaining_worth = input_parent.remaining_worth - input_parent.cost if (input_parent.remaining_worth - input_parent.cost) > 0 else 0
```
when would the remaining_worth ever be greater than the initial cost given your assumptions?

ishitap commented on ics/async_actions.py:
```diff
@@ -15,15 +19,197 @@ def update_task_descendents_flag_number(**kwargs):
+	related_tasks_with_parents_children = Task.objects.filter(is_trashed=False, pk__in=related_tasks).annotate(
+		children_list=ArrayAgg('items__inputs__task'),
+		task_parent_ids=ArrayAgg('inputs__input_item__creating_task__id'),
+		ingredients=ArrayAgg('task_ingredients__ingredient'))
```
does this get the parents of the descendants, and then the parents of those parents?

mayanb commented on ics/async_actions.py:
```diff
@@ -15,15 +19,197 @@ def update_task_descendents_flag_number(**kwargs):
+	tasks = {}
+	for x in range(0, related_tasks_with_parents_children.count()):
+		tasks[related_tasks_with_parents_children[x].id] = {'children': set(related_tasks_with_parents_children[x].children_list),
+						'process_type': related_tasks_batch_size[x].process_type_id,
```
it's not guaranteed that related_tasks_batch_size and related_tasks_with_parents_children are in the same order so i don't think we can refer to tasks across them with the same index

davidnmora committed:
Intercept PUT requests which trash tasks :hand:

davidnmora committed:
Revert "Intercept PUT requests which trash tasks :hand:"

This reverts commit 51c27af1fd4076b9d8c6d2950919fa8df4fa9bd2.

mayanb commented on ics/async_actions.py:
```diff
@@ -15,15 +19,197 @@ def update_task_descendents_flag_number(**kwargs):
+def batch_size_update(**kwargs):
+	updated_task = kwargs['pk']
+	# fetch all the descendants of updated task
+	updated_task_descendants = Task.descendants(Task.objects.filter(pk=updated_task)[0])
```
you can just do Task.objects.get(pk=updated_task).descendants()

davidnmora committed:
Update children of deleted task (@signals pre_save) :children_crossing:

addysunn committed:
add ingredient amount update

addysunn committed:
Merge branch 'update-costs' of https://github.com/polymerteam/wafflecone into update-costs

davidnmora committed:
Update parents of a deleted task (not fully tested, fails on deleted ancestors)

davidnmora committed:
Filter out trashed descendants

addysunn committed:
final commit

davidnmora committed:
Skip over trashed tasks (which throw errors when deleting tasks)

davidnmora committed:
Move helper functions to a seperate file

davidnmora committed:
Clarify misleading/incosistent names, refactor repeated batch_size code

davidnmora committed:
[BATCH SIZE DONE] Refactor parent/child $ update (lower bound @ :zero: )

davidnmora committed:
[input_update] Uncomment db saves, fix misleading comments

davidnmora committed:
Clean up/uncomment stuff in input add/delete

davidnmora committed:
Refactor shared parent ingredients code

davidnmora committed:
Comment

davidnmora committed:
Clarify variable name: total_diff -> total_change_in_value_from_all_parents

davidnmora committed:
Start tp fix update_parents_for_ingredient logic

davidnmora committed:
Clean up code

davidnmora committed:
Catch variable mistake (new_utilization instead of utilization_diff) and rename

davidnmora committed:
Decompose child update function for ingredient update

davidnmora committed:
:confetti_ball: ingredient_amount_update worksssssss!!! (at least w/ one parent)

... due to fixing the logical and arithmetic shenanigans of update_parent_for_ingredients()

davidnmora committed:
Get code working for input add and ingredient amount update (delete input untested)

davidnmora committed:
Move custom Input.delete() code to pre_delete in signals :arrow_right:

We need the delete Input signal to fire pre_delete in order to access the TaskIngredient info before it gets deleted. However, pre_delete fires before the DEFAULT delete call, not before we deleted TaskIngredient in custom code. Hence the move so we can put the cost update before we delete the TaskIngredient.

davidnmora committed:
Move all of Input Delete to pre_save :bug:

davidnmora committed:
Move Flag-related code back to Input.delete()

 (it's finicky, no need to produce a potentially new set of weirdnesses)

davidnmora committed:
Rename/refactor signal helpers

davidnmora committed:
Make input updates async compatible: pass constants where DB could change

davidnmora committed:
Clean up, delete Aditya's recipe handling code (we can do it easier now)

davidnmora committed:
[ADD/REMOVE INPUT DONE] Refactor Input.create() and consolidate create/delete fns

Now, both delete/add input assume an equal amount is coming from each parent. This means it's possible for sum_all(batch_sizes) == ingredient_amount but for not all the value of the parents to be used up, if not all parents have the same batch size.

davidnmora committed:
Make America readable again :us:

davidnmora committed:
Commit update_cost/re_cost changes, still debugging batch_size change

mayanb commented on ics/signals.py:
```diff
@@ -57,14 +56,15 @@ def item_changed(sender, instance, **kwargs):
 
 @receiver(post_save, sender=Input)
 def input_changed(sender, instance, **kwargs):
+	update_task_ingredient_for_new_input(instance)
```
this will get called every time the input is saved, regardless of whether it is added or just something about it is changed. i guess technically we aren't really editing aspects of an input right now, but i think we should be careful with this. it looks like post_save has a created boolean field so we should be using that

![image](https://user-images.githubusercontent.com/7052505/44355891-aeba5f00-a462-11e8-8143-e65ebca00f88.png)

mayanb commented on ics/signals.py:
```diff
@@ -28,27 +29,138 @@ def task_changed(sender, instance, **kwargs):
 def input_changed(sender, instance, **kwargs):
-	kwargs = { 'taskID' : instance.task.id, 'creatingTaskID' : instance.input_item.creating_task.id}
+	update_task_ingredient_for_new_input(instance)
+	kwargs = get_input_kwargs(instance, added=True)
```
here we should also be setting added by looking at the created boolean that post_save gives us

davidnmora committed:
Allow input_update to only be called when a new input is created, not updated

davidnmora committed:
Make variable names meaningful again :us:

davidnmora committed:
:confetti_ball: Re-use cost_diff fn to get batch_size update to work

davidnmora committed:
Cut trashed tasks from ArrayAggs and and parents :scissors:

davidnmora committed:
Filter trashed parents for ingredient amount changes

davidnmora committed:
Get deleted task udpate working (on basic cases)

davidnmora committed:
Get cost seeding working on toy graph (just updates single team)

davidnmora committed:
Limit to parents remaining worth

davidnmora committed:
Create Task.cost_set_by_user field which we don't update

davidnmora committed:
[BATCH SIZE WORKS!!] Create custom batch_size prev/new cost calculator

Since we assign cost predictably (first come, first serve) to children, when batch_size changes and we have to update costs, we now do it in the same order (first to last input). This is because there's the ing amount a child DESIRES to have, and the amount which the parent actually has remaining to contribute.
This is not an issue when updating the cost of something when the updated tasks batch size is unchanged. But we here had to write a custom function for batch_size changes.

davidnmora committed:
Re-design delete input so it works properly

Formerly, we just subtracted a batch_size worth of $amount. however, a deleted input may never hvae actually received any $ from the parent, so we need to "simulate" adding the input in order to figure out if it did.

davidnmora committed:
Add Maya's DB command (with reset prep added)

davidnmora commented on ics/cost_update_helpers.py:
```diff
@@ -0,0 +1,502 @@
+		update_parents_for_ingredient(parents_contributing_ingredient, old_amount, new_amount, num_parents)
+
+
+def update_parents_for_ingredient(parents_contributing_ingredient, old_amount, new_amount, num_parents, creating_task_of_changed_input=-1, input_added=False, input_deleted=False):
```
@mayanb This function, as it is here, does a ton of heavy lifting. However, I realized Friday/Monday that all the logic is unnessary: it just needs to update each ingredient alongside all of its own children, which we have a different (fairly involved) function for.

I'm stil debugging, but here's what it looks like now:

```javascript
def update_parents_for_ingredient(updated_child, parents_contributing_ingredient, old_amount, new_amount, num_parents):
	for parent_id in parents_contributing_ingredient:
		parent = parents_contributing_ingredient[parent_id]
		if parent.cost is None:
			continue
		update_creating_task_and_all_its_children(
			updated_child,
			parent_id,
			float(parent.batch_size),
			previous_total_amount_of_ingredient_for_child=old_amount,
			new_total_amount_of_ingredient_for_child=new_amount,
			input_deleted=input_deleted,
		)
```

davidnmora committed:
[ADD/REMOVE INPUTS WORKS] Iterate over parents, updating direct children

Major change: before sometimes we were JUST updating parents for an ingredient, but none of those parents children.

Now the code is simpler, for any change affecting input amount/unit_cost of an ingredient, we iterate over all parents for that ingredient, updating all their direct children (which includes the updated child).
Also, we're passing in process and product type, rather than ingredient ID.

davidnmora committed:
[BATCH SIZE, ING AMNT WORKS] Fix delete input redistribution bug
